core/shared/src/main/scala/org/http4s/UrlForm.scala

```diff
@@ -95,11 +95,25 @@ object UrlForm {
     if (values.get("").fold(false)(_.isEmpty)) new UrlForm(values - "")
     else new UrlForm(values)
 
+  def single(key: String, value: String): UrlForm =
+    new UrlForm(Map(key -> Chain.one(value)))
+
   def apply(values: (String, String)*): UrlForm =
-    values.foldLeft(empty)(_ + _)
+    values match {
+      case Seq() => empty
+      case Seq(x) => single(x._1, x._2)
+      case h +: tail => tail.foldLeft(single(h._1, h._2))(_ + _)
+    }
```

Comment:
Honestly, I am a bit concerned about this optimization. On one hand, it allows to avoid calling a lambda for sizes 0 and 1 (which is good), but on the other hand it comes at the price of three extractors called sequentially under the hood. Therefore, it may cost up to 4 intermediate objects created: 3 `Option` and then a `Tuple2` in the last `case`.

Would it be possible to give a try to an implementation based on `Iterator` as an alternative?
```scala
val it = values.iterator

if (!it.hasNext) return empty

val next = it.next()
var result = single(next._1, next._2)

while (it.hasNext) {
  result += it.next()
}

result
```
Hence, it avoids using lambdas all the way through the executions.
I haven't checked its performance though, so as for now it's just a speculative improvement.

Reply:
> Therefore, it may cost up to 4 intermediate objects created: 3 Option and then a Tuple2 in the last case.

Didn't get this. Can you please elaborate on the creation of three `Option`s in the case of a collection > 1 size for the `UrlForm.apply`?

Reply:
This was my assumption based on the fact that usually `unapply/unapplySeq` are supposed to return a result wrapped into `Option`. Actually, things turned out to be even more interesting when I dug into it deeper. Consider a simplified example:

```scala
class SeqExtractorsTest {
  def + (kv: (String, String)): SeqExtractorsTest = ???
}

object SeqExtractorsTest {

  def apply(values: (String, String)*): SeqExtractorsTest =
    values match {
      case Seq() => empty
      case Seq(x) => single(x._1, x._2)
      case h +: tail => tail.foldLeft(single(h._1, h._2))(_ + _)
    }

  def apply2(values: (String, String)*): SeqExtractorsTest = {
    val it = values.iterator
    if (!it.hasNext) return empty
    
    val next = it.next()
    var result = single(next._1, next._2)
    
    while (it.hasNext) { result += it.next() }
    
    result    
  }

  def empty: SeqExtractorsTest = ???
  def single(k: String, v: String): SeqExtractorsTest = ???
}
```
When I compile it with Scala 2.13.10, I get the following result (decompiled with CFR):
```java
public final class SeqExtractorsTest$ {
    public static final SeqExtractorsTest$ MODULE$ = new SeqExtractorsTest$();

    public SeqExtractorsTest apply(Seq<Tuple2<String, String>> values) {
        Option option;
        SeqOps seqOps;
        SeqOps seqOps2;
        Seq<Tuple2<String, String>> seq = values;
        if (seq != null && !SeqFactory.UnapplySeqWrapper$.MODULE$.isEmpty$extension(seqOps2 = package$.MODULE$.Seq().unapplySeq(seq)) && new SeqFactory.UnapplySeqWrapper(SeqFactory.UnapplySeqWrapper$.MODULE$.get$extension(seqOps2)) != null && SeqFactory.UnapplySeqWrapper$.MODULE$.lengthCompare$extension(SeqFactory.UnapplySeqWrapper$.MODULE$.get$extension(seqOps2), 0) == 0) {
            return this.empty();
        }
        if (seq != null && !SeqFactory.UnapplySeqWrapper$.MODULE$.isEmpty$extension(seqOps = package$.MODULE$.Seq().unapplySeq(seq)) && new SeqFactory.UnapplySeqWrapper(SeqFactory.UnapplySeqWrapper$.MODULE$.get$extension(seqOps)) != null && SeqFactory.UnapplySeqWrapper$.MODULE$.lengthCompare$extension(SeqFactory.UnapplySeqWrapper$.MODULE$.get$extension(seqOps), 1) == 0) {
            Tuple2 x = (Tuple2)SeqFactory.UnapplySeqWrapper$.MODULE$.apply$extension(SeqFactory.UnapplySeqWrapper$.MODULE$.get$extension(seqOps), 0);
            return this.single((String)x._1(), (String)x._2());
        }
        if (seq != null && !(option = package$.MODULE$.$plus$colon().unapply(seq)).isEmpty()) {
            Tuple2 h = (Tuple2)((Tuple2)option.get())._1();
            Seq tail = (Seq)((Tuple2)option.get())._2();
            return (SeqExtractorsTest)tail.foldLeft((Object)this.single((String)h._1(), (String)h._2()), (Function2 & Serializable)(x$1, x$2) -> x$1.$plus((Tuple2<String, String>)x$2));
        }
        throw new MatchError(seq);
    }

    public SeqExtractorsTest apply2(Seq<Tuple2<String, String>> values) {
        Iterator it = values.iterator();
        if (!it.hasNext()) {
            return this.empty();
        }
        Tuple2 next = (Tuple2)it.next();
        SeqExtractorsTest result = this.single((String)next._1(), (String)next._2());
        while (it.hasNext()) {
            result = result.$plus((Tuple2<String, String>)((Tuple2)it.next()));
        }
        return result;
    }

    public SeqExtractorsTest empty() {
        throw Predef$.MODULE$.$qmark$qmark$qmark();
    }

    public SeqExtractorsTest single(String k, String v) {
        throw Predef$.MODULE$.$qmark$qmark$qmark();
    }

    private SeqExtractorsTest$() {
    }
}
```

Reply:
Thus, it looks like `Seq#unapplySeq` pretends to be optimized and does not cause any intermediate `Option`, but creates an instance of `SeqFactory.UnapplySeqWrapper` instead (despite that it extends `AnyVal`).

Furthermore, the `unapply` for `+:` extractor still creates an intermediate `Option` along with `Tuple2` for the sake of the extraction only.

Reply:
Also I'm getting this compiler warning with `-Xlint` option enabled:
```
../SeqExtractorsTest.scala:9: warning: match may not be exhaustive.
It would fail on the following input: (x: Seq[?] forSome x not in Nil)
    values match {
    ^
1 warning
```
Apparently, the compiler cannot make a connection that the `+:` extractors covers all the cases for size > 1.

Reply:
Needless to mention, the implementation based on an iterator compiles into way cleaner and concise byte code almost 1-to-1 as it is written in Scala without any sudden indirections nor instance creations.

Reply:
@satorg wow, I somehow forgot about your comments on this. Do your comments are blockers for getting this merged?

Reply:
https://www.scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#extractor-patterns

> A type is said to be an extractor type for some type T if it has a method get with return type T, and a method isEmpty with a return type that conforms to Boolean. Option[T] is an extractor type for type T.